packs/orion/actions/lib/actions.py

```diff
@@ -0,0 +1,212 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+from st2actions.runners.pythonrunner import Action
+from orionsdk import SwisClient
+
+
+class OrionBaseAction(Action):
+    def __init__(self, config):
+        super(OrionBaseAction, self).__init__(config)
+
+        self.client = None
+
+        if "orion" not in self.config:
+            raise ValueError("Orion host details not in the config.yaml")
+
+    def connect(self, platform):
+        """
+        Connect to an Orion platform from the packs config.yaml.
+        """
+        try:
+            self.client = SwisClient(
+                self.config['orion'][platform]['host'],
+                self.config['orion'][platform]['user'],
+                self.config['orion'][platform]['password'])
+        except KeyError:
+            raise ValueError("Orion host details not in the config.yaml")
+
+    def query(self, swql, **kargs):
+        """
+        Run SWQL against the Orion Platform.
+        """
+        return self.client.query(swql, **kargs)
+
+    def invoke(self, entity, verb, *args):
+        """
+        Run an Invoke against the Orion Platform.
+        """
+        return self.client.invoke(entity, verb, *args)
+
+    def create(self, entity, **kargs):
+        """
+        Run an Create against the Orion Platform.
+        """
+        return self.client.create(entity, **kargs)
+
+    def node_exists(self, caption, ip_address):
+        """
+        Check if an Node exists (caption and or ip) on the Orion platform.
+
+        Returns: True or False.
+        """
+        swql = """SELECT NodeID, IPAddress FROM Orion.Nodes
+                  WHERE Caption=@caption"""
+        kargs = {'caption': caption}
+        caption_data = self.query(swql, **kargs)
+
+        swql = """SELECT NodeID, IPAddress FROM Orion.Nodes
+                  WHERE IPAddress=@ip_address"""
+        kargs = {'ip_address': ip_address}
+        ip_data = self.query(swql, **kargs)
+
+        if len(caption_data['results']) >= 1 or len(ip_data['results']) >= 1:
+            return True
+        else:
+            return False
+
+    def get_node_id(self, caption):
+        """
+        Gets an NodeID from the Orion platform.
+
+        Raises: ValueError on muliple or no matching caption.
+
+        Returns: the NodeID (int)
+        """
+        swql = "SELECT NodeID FROM Orion.Nodes WHERE Caption=@caption"
+        kargs = {'caption': caption}
+        data = self.query(swql, **kargs)
+
+        if len(data['results']) == 1:
+            try:
+                return data['results'][0]['NodeID']
+            except IndexError:
```

Comment:
Might be a good idea to debug log result on failure to make it easier to debug those scenarios (e.g. is the array empty or the first item in the array is missing NodeId key...)